src/components/Todo/TodoItem.tsx

```diff
@@ -0,0 +1,169 @@
+import { Todo } from '../../types/Todo';
+import cn from 'classnames';
+import { useTodosContext } from '../../utils/useTodosContext';
+import { Errors } from '../../types/Errors';
+import * as todoSevice from '../../api/todos';
+import { handleRequestError } from '../../utils/handleRequestError';
+import { useEffect, useRef, useState } from 'react';
+
+type Props = {
+  todo: Todo;
+  deleteTodo: (id: number) => void;
+};
+
+export const TodoItem: React.FC<Props> = ({ todo, deleteTodo }) => {
+  const {
+    loadingTodoIds,
+    setError,
+    setTodos,
+    onDelete,
+    todos,
+    setIsFocused,
+    setLoadingTodoIds,
+  } = useTodosContext();
+  const [isEdit, setIsEdit] = useState(false);
+  const [changeTitle, setChangeTitle] = useState(todo.title);
+
+  const changeInput = useRef<HTMLInputElement>(null);
+
+  useEffect(() => {
+    if (isEdit && changeInput.current) {
+      changeInput.current.focus();
+      setIsFocused(false);
+    }
+  }, [isEdit]);
+
+  const updateTodo = () => {
+    setError(Errors.default);
+    setLoadingTodoIds(prev => [...prev, todo.id]);
+
+    if (changeTitle === todo.title) {
+      setIsEdit(false);
+      setLoadingTodoIds([]);
+
+      return;
+    }
+
+    if (!changeTitle.trim().length) {
+      return onDelete(todo.id);
+    }
+
+    return todoSevice
+      .updateTodo(todo.id, { title: changeTitle.trim() })
+      .then(todoItem => {
+        setTodos(currentTodos => {
+          const newTodos = [...currentTodos];
+          const index = newTodos.findIndex(t => t.id === todo.id);
+
+          newTodos.splice(index, 1, todoItem);
```

Comment:
- Use `map` instead of `findIndex` + `splice`.
- Avoid one-letter variables/arguments